### PR TITLE
Fix Server.cleanup_jobs in PTL

### PR DESCRIPTION
**Pull request by hirenvadalia**
#### Describe Bug or Feature
* Currently, Server._cleanup_large_num_jobs() finds session_id of job and it runs `kill -9 pid` on server node but in case of multinode setup this not valid as the job can run on mom node and session_id will be from that mom node not from the server node
* When running qdel/pbs_rdel command with large num of job/resv, qdel/pbs_rdel can fail with `Argument too long` error as there are limits on max args passed to any command

#### Describe Your Change
* Remove buggy Server._cleanup_large_num_jobs() and force cleanup_jobs to do qdel -Wforce as root user so Server won't wait for Mom's reply of deleing job
* In Sever.deljob and Server.delresv, if given ids are more than 2k then run multiple qdel/pbs_rdel with max 2k id at a time
* Add `nomail` in convert_to_cli for IFL_DELETE operation to suppress mails on qdel
* Fix below issues with Travis scripts:
    * Use absolute path for PTL reqs file.
    * change the permission of pbs_iff and pbs_rcp after make install without any condition

#### Link to Design Doc
None

#### Attach Test and Valgrind Logs/Output
Not required as cleanup_jobs is used in Server.reverts_to_defaults so Travis will be sufficient

### CODE REVIEW

**Comment by arungrover** [comment]
I suggest you run performance test suite with this change and notice the amount of time cleanup takes across testcases. I suspect the reason we were killing session IDs was because qdel is extremely slow (with or without force). Just killing the pids make mom catch sigchild and cleanup quickly.

**Comment by hirenvadalia** [comment]
@arungrover Thanks for asking to run tests... While I was doing changes I measured cleaup_jobs performance using `timeit` module but didn't realize that we have a restart of mom in setUp which will get slow down...

> I suggest you run performance test suite with this change and notice the amount of time cleanup takes across testcases.

Here are logs:
[before_fix.txt](https://github.com/PBSPro/pbspro/files/3554595/before_fix.txt)
[after_fix.txt](https://github.com/PBSPro/pbspro/files/3554598/after_fix.txt)
```
hirenk@blrlap821:~/code/pbspro$ grep -n 'Tests run in' before_fix.txt after_fix.txt 
before_fix.txt:186759:Tests run in 0:53:03.974624
after_fix.txt:88266:Tests run in 0:51:23.645206
hirenk@blrlap821:~/code/pbspro$ 
```
New changes don't slow down cleanup of the job but it slows down restart of mom since mom is catching up with sigchild on job's processes.
I will look into killing session_id correct way (aka run kill command only on MS) and see if that helps in a faster restart of mom or not.


> because qdel is extremely slow (with or without force)

Agree... but there is one case where qdel is fast, and that is if the user is root or PBS mgr user and scheduling is off. which we are using in cleanup_jobs to make qdel fast

**Comment by hirenvadalia** [comment]
Marking this PR WIP based on above comment...

**Comment by hirenvadalia** [comment]
[new_changes_try1.txt](https://github.com/PBSPro/pbspro/files/3559158/new_changes_try1.txt)
[new_changes_try2.txt](https://github.com/PBSPro/pbspro/files/3559160/new_changes_try2.txt)

File: test/tests/functional/pbs_offline_vnodes.py [hunk 1]
```diff
@@ -1,490 +1,490 @@
...
-        self.server.expect(
-            NODE, {ATTR_NODE_state: 'offline'},
-            id=momB.shortname, interval=2)
-
-        # momA and the vnodes she reports should be free
-        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
-                           id=momA.shortname, interval=2)
-
-        self.verify_vnodes_state('free')
+# coding: utf-8
```
Comment (arungrover) [inline test/tests/functional/pbs_offline_vnodes.py]: why is this whole file showing up as changed? what line did you change in this file?

File: test/fw/ptl/lib/pbs_testlib.py [hunk 2]
```diff
@@ -8411,108 +8436,102 @@ def is_history_enabled(self):
             return True
         return False
 
-    def cleanup_jobs(self, extend=None, runas=None):
+    def cleanup_jobs(self):
```
Comment (arungrover) [inline test/fw/ptl/lib/pbs_testlib.py]: why did you remove extend and runas fields from this function? What if some test wants to do deletehistory as part of cleaning up jobs?

File: test/fw/ptl/lib/pbs_testlib.py [hunk 3]
```diff
@@ -6224,13 +6224,26 @@ def deljob(self, id=None, extend=None, runas=None, wait=False,
...
                 as_script = False
-            ret = self.du.run_cmd(self.client, pcmd, runas=runas,
-                                  as_script=as_script, logerr=logerr,
-                                  level=logging.INFOCLI)
-            rc = ret['rc']
-            if ret['err'] != ['']:
-                self.last_error = ret['err']
-            self.last_rc = rc
+            if id is not None:
+                chunks = [id[i:i + 2000] for i in range(0, len(id), 2000)]
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: > How did you come up with number 2000?

I didn't do much calculation or thinking on this, I just decided to use 2k based on the value I get from `(((getconf ARG_MAX) - 1000)/PBS_MAXSVRJOBID)/2)` on my system

> isn't it dependent on system configuration? So on some systems even 2000 could be big enough that the commands can not take it.

Yes, it is but I don't think so there will be system which can't take ~54k (~2k jobid) character on the command line.
Also, I don't to waste time here in doing unnecessary caclution on chunk size, if we find any problem in 2k then will reduce it in the future...

File: .travis/do.sh [hunk 4]
```diff
@@ -1,9 +1,19 @@
 #!/bin/bash -xe
 
+if [ $(id -u) -ne 0 ]; then
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply .travis/do.sh]: Yes, this change is not related to this PR but I just included since while doing testing of this PR I found few issued in Travis script.

File: test/fw/ptl/lib/pbs_testlib.py [hunk 5]
```diff
@@ -2429,10 +2429,11 @@ def convert_to_cli(self, attrs, op=None, hostname=None, dflt_conf=True,
             if isinstance(attrs, list):
                 for a in attrs:
                     if 'force' in a:
-                        _c.append('-W')
-                        _c.append('force')
+                        _c.append('-Wforce')
                     if 'deletehist' in a:
                         _c.append('-x')
+                    if 'nomail' in a:
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: it is part of this PR, see cleanup_jobs() it is using nomail while invoking qdel

File: test/fw/ptl/lib/pbs_testlib.py [hunk 6]
```diff
@@ -8411,108 +8436,102 @@ def is_history_enabled(self):
...
         select_xt = None
         if self.is_history_enabled():
             delete_xt += 'deletehist'
             select_xt = 'x'
-        job_ids = self.select(extend=select_xt)
+        jobs = self.status(JOB, extend=select_xt)
+        job_ids = sorted(list(set([x['id'] for x in jobs])))
+        host_pid_map = {}
+        for job in jobs:
+            if 'exec_host2' in job:
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: While I was looking into code to confirm that first node name in exec host will MS or not, I find almost all code is looking at JOB_ATR_exec_host2 or ATTR_exechost2 so I thought it will be visible to use so I used, but if you are saying it's not visible then I will remove it.

File: test/fw/ptl/lib/pbs_testlib.py [hunk 7]
```diff
@@ -8411,108 +8436,102 @@ def is_history_enabled(self):
...
+            if sc['scheduling'] == 'True':
+                sched_state.append(sc['id'])
+                # runas is required here because some tests remove
+                # current user from managers list
+                a = {'scheduling': 'False'}
+                self.manager(MGR_CMD_SET, SCHED, a, id=sc['id'],
+                             runas='root')
+                self.expect(SCHED, a, id=sc['id'])
+        try:
+            self.deljob(id=job_ids, extend=delete_xt, runas='root', wait=False)
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: If we do kill session-id then the server will get obit storm, also here we are doing qdel with Wforce as a root user so the server will not contact to mom at all.

File: test/fw/ptl/lib/pbs_testlib.py [hunk 8]
```diff
@@ -8411,108 +8436,102 @@ def is_history_enabled(self):
...
+        except PbsDeljobError:
+            pass
+        st = int(time.time())
+        if len(job_ids) > 100:
+            for host, pids in host_pid_map.items():
+                pids = list(set(pids))
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: ok will remove it.

File: test/fw/ptl/lib/pbs_testlib.py [hunk 9]
```diff
@@ -8411,108 +8436,102 @@ def is_history_enabled(self):
...
         """
-        delete_xt = 'force'
+        delete_xt = 'nomailforce'
         select_xt = None
         if self.is_history_enabled():
             delete_xt += 'deletehist'
             select_xt = 'x'
-        job_ids = self.select(extend=select_xt)
+        jobs = self.status(JOB, extend=select_xt)
+        job_ids = sorted(list(set([x['id'] for x in jobs])))
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: because I need find last job id for log_match

File: test/fw/ptl/lib/pbs_testlib.py [hunk 10]
```diff
@@ -8411,108 +8436,102 @@ def is_history_enabled(self):
...
+                self.expect(SCHED, a, id=sc['id'])
+        try:
+            self.deljob(id=job_ids, extend=delete_xt, runas='root', wait=False)
+        except PbsDeljobError:
+            pass
+        st = int(time.time())
+        if len(job_ids) > 100:
+            for host, pids in host_pid_map.items():
+                pids = list(set(pids))
+                chunks = [pids[i:i + 5000] for i in range(0, len(pids), 5000)]
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: same answer as 2k job id.

File: test/fw/ptl/lib/pbs_testlib.py [hunk 11]
```diff
@@ -8411,108 +8436,102 @@ def is_history_enabled(self):
...
+                for chunk in chunks:
+                    self.du.run_cmd(host, ['kill', '-9'] + chunk,
+                                    runas='root', logerr=False)
+            _msg = job_ids[-1] + ';'
+            _msg += 'Job Obit notice received has error 15001'
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: As I said in comment in expect block we need to give some time to mom to catch with sigchild else in next test when mom restarts it will take time and that may cause problem for next test (like timeout of test or failure of start of mom because init.d only waits for 180s max for mom to shutdown gracefully)

File: test/tests/functional/pbs_offline_vnodes.py (hunk 1 above)
Comment (hirenvadalia) [reply test/tests/functional/pbs_offline_vnodes.py]: previously file had CRLF line ending when I changed file my editor converted all line endings to LF. I changed line no. 133, removed args from cleanup_jobs()

File: test/fw/ptl/lib/pbs_testlib.py (hunk 2 above)
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: > why did you remove extend and runas fields from this function?

It was useless after new changes, in this PR we made cleanup_jobs to clean jobs at root user with nomail, force, deletehistory (if applicable).



> What if some test wants to do deletehistory as part of cleaning up jobs?

cleanup_jobs was never used extend passed to it (even before changes to this PR) and now also not using since it is passing its own extend (nomail, force, deletehistory) to select() and deljob().

File: test/fw/ptl/lib/pbs_testlib.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arungrover) [reply test/fw/ptl/lib/pbs_testlib.py]: Fair enough, it was important to know how you pulled out this number

File: test/fw/ptl/lib/pbs_testlib.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (arungrover) [reply test/fw/ptl/lib/pbs_testlib.py]: I don't think using Wforce option means that the server will not contact mom at all. It means that the server will send a deletejob request to mom and then delete jobs without waiting for an Obit from mom. 
I see your point that killing session ids might trigger an obit storm and it could potentially be slower.

File: test/fw/ptl/lib/pbs_testlib.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (arungrover) [reply test/fw/ptl/lib/pbs_testlib.py]: I understand, but before your change, we had no check for log match and we did not see any problem with timeout or failure to start mom because of init script timing out. 
Also, code on line 8502 is doing exactly similar to what you are trying to do. It is counting job_state attributes to sum up to 0. I'd suggest you retain either the log msg check or keep expect on job_state.

File: test/fw/ptl/lib/pbs_testlib.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: @arungrover Sorry my mistake I didn't mean "contact" but I mean "wait for mom's reply on delete job"

File: test/fw/ptl/lib/pbs_testlib.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: @arungrover before changes to this PR it was deleting job one by one (aka one qdel per jid) + kill session_id + it was doing expect() for the queue attr to unset on per job (aka qstat per job). So cleanup_jobs was invoking 3 commands (qdel + kill + qstat) per jid and then waiting for job_stat to be 0 so it was too slow that we didn't have to wait for mom to process sigchild.
But after this PR changes we made qdel and kill command much faster + we removed qstat call as it's useless because with qdel -Wforce (as root user) server will immediately delete a job from itself so there is no point of waiting for the queue attr to unset on the job. So now we need a new check which will wait and allow some time to mom to process sigchild which is log match and then after just to ensure we do check job_state to be 0 on the server.

**Comment by hirenvadalia** [comment]
> Didn't you say that you will remove making the list of pids a set? I do not see that change.
> I also have a small comment about the log_match

Sorry, I missed removing set on pids, will do it now. Also, I will update to use `ROOT_USER` instead hardcoded `root` since it is available now in pbs_testlib

File: test/fw/ptl/lib/pbs_testlib.py [hunk 12]
```diff
@@ -8293,108 +8318,99 @@ def is_history_enabled(self):
...
+        while reservations:
             resvs = [r['id'] for r in reservations]
             if len(resvs) > 0:
                 try:
-                    self.delresv(resvs, logerr=False, runas=runas)
+                    self.delresv(resvs, runas=ROOT_USER)
                 except:
                     pass
-                reservations = self.status(RESV, level=logging.DEBUG)
+                reservations = self.status(RESV)
```
Comment (kjakkali) [inline test/fw/ptl/lib/pbs_testlib.py]: add runas user here

Comment (hirenvadalia) [reply test/fw/ptl/lib/pbs_testlib.py]: done

**Comment by hirenvadalia** [comment]
Had to do 2nd force-push to resolve conflicts...
